### ISSUE TITLE
AttributeError: module 'numpy' has no attribute 'bool'.

### ISSUE THREAD

**Issue by Glfrey** [mhalushka/miRge3.0#63]
Hi @arunhpatil,

I've been successful running Mirge3.0 on a ubuntu server. However, I had to set up a new one and I'm now getting the following error when calling Mirge3.0:

```
(base) $ miRge3.0
/home/ubuntu/.local/lib/python3.8/site-packages/pandas/_testing.py:24: FutureWarning: In the future `np.bool` will be defined as the corresponding NumPy scalar.
  import pandas._libs.testing as _testing
Traceback (most recent call last):
  File "/home/ubuntu/.local/bin/miRge3.0", line 5, in <module>
    from mirge.__main__ import main
  File "/home/ubuntu/.local/lib/python3.8/site-packages/mirge/__main__.py", line 12, in <module>
    import pandas as pd
  File "/home/ubuntu/.local/lib/python3.8/site-packages/pandas/__init__.py", line 180, in <module>
    import pandas.testing
  File "/home/ubuntu/.local/lib/python3.8/site-packages/pandas/testing.py", line 5, in <module>
    from pandas._testing import (
  File "/home/ubuntu/.local/lib/python3.8/site-packages/pandas/_testing.py", line 24, in <module>
    import pandas._libs.testing as _testing
  File "pandas/_libs/testing.pyx", line 10, in init pandas._libs.testing
  File "/home/ubuntu/.local/lib/python3.8/site-packages/numpy/__init__.py", line 305, in __getattr__
    raise AttributeError(__former_attrs__[attr])
AttributeError: module 'numpy' has no attribute 'bool'.
`np.bool` was a deprecated alias for the builtin `bool`. To avoid this error in existing code, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
```

I'll do my own investigation with regards to what is making the two set ups different, but as this seems to be Numpy version-related error I thought best to share with you.

Best wishes,

Gill

**Comment by Glfrey** [comment]
Hi @arunhpatil ,

I fixed the error by downgrading numpy:

```
(base) $ python3.8 -m pip install numpy==1.19
Defaulting to user installation because normal site-packages is not writeable
Collecting numpy==1.19
  Downloading numpy-1.19.0-cp38-cp38-manylinux2010_x86_64.whl (14.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 14.6/14.6 MB 38.2 MB/s eta 0:00:00
Installing collected packages: numpy
  Attempting uninstall: numpy
    Found existing installation: numpy 1.24.2
    Uninstalling numpy-1.24.2:
      Successfully uninstalled numpy-1.24.2
Successfully installed numpy-1.19.0
(base) $ miRge3.0
usage: miRge3.0 [options]

miRge3.0 (Comprehensive analysis of small RNA sequencing Data)

optional arguments:
  -h, --help  show this help message and exit
```
Probably worth a mention on your installation instructions that numpy version needs to be < 1.20 if you get the chance.

Best wishes,

Gill

**Comment by arunhpatil** [comment]
Hi @Glfrey,

I will check and correct this based on what version of numpy is installed in the machine. I think this should be fixed before more users come up with this issue. 
`AttributeError: module 'numpy' has no attribute 'bool'.`

Thank you again, I will replace every occurance of a `np.bool` attribute to `bool`, so that we can keep up with the growing numpy package. 

I will update once I have this fix and then we can close this issue.

Thank you,
Arun.

**Comment by arunhpatil** [comment]
Hi @Glfrey,

I was updating miRge3.0 and was examining this particular issue. The numpy I have is of higher version (but not current, which is 1.25.0), and same is the case with pandas (current 1.5.3). `The error refers to not call np.bool but instead to use bool` (more about this error at the bottom). I checked all existing scripts, there are only three lines of code that use np.int32 and np.float32 (listed below). I performed a test run by calling the function A-to-I editing (which uses thoses int/float) and there was no error.

Versions on my machine:
```
>>> import numpy as np
>>> np.__version__
'1.22.0'
>>> import pandas as pd
>>> pd.__version__
'1.4.3'
>>>
```

Script that specify numpy datatypes: 
```
mirge2_tRF_a2i.py:    return np.array(rho, np.float32)
mirge2_tRF_a2i.py:    return np.array(delta, np.float32), np.array(nneigh, np.int32), np.array(nneighDistance, np.float32), sort_rho_idx.astype(np.int32)
mirge2_tRF_a2i.py:            cl = cl.astype(np.int32)
```

The error is with older version of padas, as you see in line six (`import pandas.testing`):
```
File "/home/ubuntu/.local/bin/miRge3.0", line 5, in <module>
    from mirge.__main__ import main
  File "/home/ubuntu/.local/lib/python3.8/site-packages/mirge/__main__.py", line 12, in <module>
    import pandas as pd
  File "/home/ubuntu/.local/lib/python3.8/site-packages/pandas/__init__.py", line 180, in <module>
    import pandas.testing
```
The pandas in this case seems to be of older version, which requires older version of numpy.  When pandas.testing was called, the downstream function had np.bool, which is depricated in the latest versions of numpy. We could either install lower version of numpy or update the pandas package.  (I suspect python version to be <3.7 and this could be why we see older version of pandas)

I hope I am not wrong in debugging this, and I will refrain from making any changes to the code unless this issue recur. 

Thank you,
Arun.
